Собираются 4 разных дистрибутивов под NET-CORE 3.1

aliczin opened the pull request:
через отладочный файл publish.bat

![изображение](https://user-images.githubusercontent.com/56105699/97707075-c9b79d80-1ac7-11eb-82b7-3aee4b5cd2f6.png)

из интересного 

* включен JIT
* весь Runtime убран в каталог `RUNTIME` через библиотеку https://github.com/nulastudio/NetCoreBeauty#why-and-why-not
* сделал каталог ресурсов внутри проекта OScript - откуда в момент публикации они копируются (обновлен OPM до текущего релиза в этих самых ресурсах)

Осталось

- [X] Сделать одну универсальную команду публикации вместе с RUNTIME через `dotnet pubish`
- [ ] **Добится прохождения тестов на всех платформах в режиме JIT - через штатный testrunner - для этого методы портированы в виде пул-реквестов в https://github.com/oscript-library/asserts/pull/9, и в https://github.com/oscript-library/1testrunner/pull/16**
- [ ] **Установить стандартный комплект библиотек через OPM**
- [x] Сделать "Зиппер" финальных файликов под платформы
- [x] Разобраться  почему NET 5.0 берет версию не оттуда откуда берет NETCore 3.1 - а лучше всего передавать версию единооборазно
- [x] Почить AST и глобальные методы #1039 
- [x] Привести трассировку загрузки библиотек к формату разных локалей -EN/RU
- [x] Проанализировать WARNING JIT компилятора под Windows
- [x] Попробовать оставить `lib.system = ../lib` сделав новый класс поиска полных путей
- [x] Вернуть старое поведение дефолтного package-loader - на отсутствие перебора через рекурсию

Дополнительно - уже на всякий случай для целей тестирования, в качестве НЕ основной платформы добавлен NET 5.0, чтобы можно было "матрично тестировать" 

#1037 

работа будет продолжена - но @EvilBeaver уже можешь посмотреть на хитрости с флагами в dotnet publish

aliczin committed:
Собираются 8 разных дистрибутивов
через отладочный файл publish.bat

aliczin committed:
Не удалять дитрибутивы для 5-ки
запуск тестов - пока падают

aliczin committed:
Пока не грузится последний OPM
добавлена трассировка загрукзи библиотек - давно было пора
поменял генерацию GLOBAL.json

aliczin committed:
В NetCore меняется поиск текущего Assembly
приходится искать в трех местах

aliczin committed:
Добавил трассировку откуда файла настроек

aliczin committed:
У нас появился каталог runtime
поэтому штатные библиотеки выше уровнем

aliczin committed:
Теперь падает JSON из-за не штатного
package-loader

aliczin committed:
Явно выделил что работает новые AST Compiler

aliczin committed:
Работает архивация артефакатов

aliczin committed:
Финальный подход - работает публикация
падает AST

nixel2007 commented on src/ScriptEngine.HostedScript/LibraryLoader.cs:
```diff
@@ -214,14 +219,17 @@ private bool CustomizedProcessing(string libraryPath)
         private bool DefaultProcessing(string libraryPath)
         {
-            var files = Directory.EnumerateFiles(libraryPath, "*.os")
+            var files = Directory.EnumerateFiles(libraryPath, "*.os", SearchOption.AllDirectories)
```
смена поведения

aliczin replied:
@nixel2007 да скорее ты прав

![изображение](https://user-images.githubusercontent.com/56105699/97783586-e0362580-1ba9-11eb-8429-d59970fad0df.png)

в package-loader используется
![изображение](https://user-images.githubusercontent.com/56105699/97783599-f348f580-1ba9-11eb-92d1-3c17e70af24b.png)

уберем

nixel2007 commented on src/ScriptEngine.HostedScript/oscript.cfg:
```diff
@@ -1,7 +1,7 @@
 
 # Корневой каталог системных библиотек
-lib.system = ../lib
+lib.system = ../../lib
```
если oscript.cfg все еще рядом с oscript.exe, зачем переносить каталог?
тем более это может быть breaking change для библиотек, оперирующих путями библиотек

aliczin replied:
А тут сложней - формально у нас ExecutionAssembly - ScriptingEndgine, а не oscript.exe.
То есть всегда так было - корневой каталог искался от DLL - а не относительно oscript.exe.

что делать пока неясно - скорее всего делать специальный класс PathUtils который будет искать всякое.
В 5-рке вообще поменяли все методы поиска "ИспольняемогоАссембли"

пока падает AST - после этого этим займусь

EvilBeaver replied:
@aliczin договорились же (и фиксировали) что чиним запуск на 3.1 а не на пятерке. Может закроем вопрос с 3.1 где не менялись ExecutionAssembly? А пятеркой заниматься отдельно, а?

aliczin replied:
@EvilBeaver падает именно на тройке - 5-ку я не трогал даже - просто добавил вторым фреймворком - щас уберу

aliczin committed:
вернул старое поведение

aliczin committed:
Пример перевода на локаль
трассировки загрузки библиотек
дальше будет в таком же стиле

aliczin committed:
Дальнейший рефакторинг трасировки

aliczin committed:
Продолжение трассировки
AST

aliczin committed:
Добавил в стек
номера конфликтов методов

aliczin committed:
Приведение трассировки к единому стилю
вообще конечно еще бы и Exception's

aliczin committed:
Убрал 5.0
Убрал отдельный каталог
чтобы четко воспроизвести ошибку AST

aliczin committed:
Сделал просто PUBLISH
одной версии
без изменений путей
без оптмизации
без триминга
AST падает

aliczin commented on src/oscript/oscript.csproj:
```diff
@@ -3,7 +3,8 @@
   <PropertyGroup>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <TargetFrameworks>netcoreapp3.1</TargetFrameworks>
+    <!--<BeautyLibsDir>runtime</BeautyLibsDir>-->
```
Оставлено на будущее

aliczin commented on src/oscript/oscript.csproj:
```diff
@@ -42,4 +44,81 @@
 
-</Project>
\ No newline at end of file
+  <ItemGroup>
```
@EvilBeaver отсюда начинается непосредственно PUBLISH скрипт

aliczin commented on src/oscript/oscript.csproj:
```diff
@@ -42,4 +44,81 @@
+
+  <Target AfterTargets="Publish" Name="InstallStandartLibrary">
+    <Message Importance="High" Text="Installing standart library directly from hub.oscript"/>
+    <Exec Command="echo ending publish" />
```
@EvilBeaver когда OPM заработает - здесь можно добавить установку пакетов прямо с HUB при помощи OPM

aliczin committed:
Локально проверяемый
полнофункциональный скрипт публикации - выдает 4 архива с движком
проверяет тесты

aliczin committed:
ДОбавил все таки оптимизацию под Windows

aliczin commented:
Финальное поведение - что в итоге на сейчас

можно собирать даже через GitHub Action
```
dotnet restore .\src\oscript\oscript.csproj
dotnet build .\src\oscript\oscript.csproj
```

а публиковать через `matrix.os`

команда для публикации сделана в `publish.bat` - все 4 архива

но можно вызывать и на Jenkins/GitHub-Actions

пример


> dotnet publish -p:PublishTrimmed=true -f netcoreapp3.1 -p:DebugType=None -r osx-x64  -c Release --self-contained true --force -p:IncludeNativeLibrariesInSingleFile=true -p:CopyOutputSymbolsToPublishDirectory=false .\src\oscript\oscript.csproj -o distrs\net31\osx-x64\bin

появится и движок и артефакт для загрузки на GitHub например

EvilBeaver commented:
@aliczin выпили, будь добр, из пулреквеста тестовый код по ошибке #1039 Она исправлена коммитом d6bd5d0, нужно разрешить конфликт

aliczin committed:
Merge branch 'feature/compiler-v2' of https://github.com/EvilBeaver/OneScript into feature/compiler-v2-publisher

aliczin commented:
@EvilBeaver `DONE`

EvilBeaver commented:
Предлагаю пункт "Добится прохождения тестов на всех платформах в режиме JIT" не делать, т.к. это за скоупом задачи "сделать билд-скрипт". Достаточно просто пометить шаг запуска тестов метками "закомментировать от сих до сих" чтобы не запускался заведомо падающий шаг

EvilBeaver commented on src/ScriptEngine.HostedScript/EngineConfigProvider.cs:
```diff
@@ -35,12 +35,22 @@ public static string DefaultConfigFilePath()
             var configFile = System.IO.Path.Combine(assemblyPath, CONFIG_FILE_NAME);
-            if (System.IO.File.Exists(configFile))
+            if (System.IO.File.Exists(configFile)) {
+                LibraryResolver.TraceLoadLibrary(String.Format("Файл настроек по умолчанию для загрузки библиотек, расположен {0}", configFile));
```
Вот эти улучшения трейса лучше отделить от скрипта сборки под netcore и отправить PR на эту функциональность в основной движ на net4

EvilBeaver replied:
А почему resolved?

aliczin replied:
упс

aliczin committed:
Merge branch 'feature/compiler-v2' into feature/compiler-v2-publisher

aliczin committed:
Оповещение

aliczin committed:
Merge branch 'feature/compiler-v2' of https://github.com/EvilBeaver/OneScript into feature/compiler-v2-publisher

aliczin committed:
Борьба за трассировку
opm не определяется как источник - поэтому
поиск идет в каталоге RUNTIME

aliczin committed:
Еще трасировочка
пора писать тесты ;-)

aliczin committed:
Жуть

aliczin commented:
Борьба продолжается

![изображение](https://user-images.githubusercontent.com/56105699/101493386-1029d300-3977-11eb-9fa3-7176c0953d45.png)

Почему-то источник вызова пустой - после чего относительный каталог берется ОТ RUNTIME
по моему это именно что новое поведение .NetCore с путями
@EvilBeaver подскажи - а каком пакете тесты на загрузчик библиотек - трассировать уже устал - надо тесты пописать

aliczin commented:
@nixel2007 - это не то.

aliczin committed:
Merge branch 'feature/compiler-v2' of https://github.com/EvilBeaver/OneScript into feature/compiler-v2-publisher

aliczin committed:
починил Clean
убрал для отладки текущего пути скрипта

aliczin committed:
Правильный отлоадчный скрипт публикации

aliczin committed:
Совокупно 8 тестов падают
чего говорить про opm

aliczin committed:
Падающий тест загрузчика пакетов

aliczin commented:
Собственно падающий тест повторяющий на чем падают opm и notify и fluent

![изображение](https://user-images.githubusercontent.com/56105699/101525138-f0f16c80-399b-11eb-8f87-7f9ab242943f.png)

Почему-то в использования метода `Использовать` - объект Source имеет `Null` - соответственно поиск идет относительно каталога запуска - применительно к запускателю тестов это каталог тесты. А должен быть каталог главного модуля

EvilBeaver commented:
@aliczin понятно, тогда опять ставим на hold и ждем пока я переделаю загрузчик. Однако, не очень понятно, потому что здесь в одном PR и сборка под дистрибутивы и тестирование сторонних приложений. Поэтому я и предлагал разделить скрипты сборки и вообще валидацию работы всего что есть под netcore.

Т.е. в моей картине мира есть 2 реквеста. 

- Первый билдит даже потенциально неработающий 1скрипт. Тесты самого движка сейчас проходят, этого достаточно для создания дистрибутива
- Второй - доработки запуска под net5

А тут получается эти задачи связаны в одном PR и не очень понятно как разделить две задачи, первая явно краткосрочная и простая, а вторая - долгая и мучительная.

aliczin commented:
@EvilBeaver уже начал разделять - этот PR закрою - будет

* feature/library-resolver-tracer - в develop
* feature/distrs-publisher - в compiler-v2

а дальше уже продолжим, чтобы opm ставил нужные версии библиотек сразу в дистрибутив

aliczin commented:
https://github.com/EvilBeaver/OneScript/pull/1053

aliczin commented:
https://github.com/EvilBeaver/OneScript/pull/1054
